# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/gtk/WaylandVsyncSource.h
# Commit: 3c74d415fafd
# Full Hash: 3c74d415fafd4bff5576ed0a2256c1a0232b5937
# Author: Robert Mader <robert.mader@posteo.de>
# Date: 2021-06-11 09:42:05
# Regressor Bug: 1711244
# File Overlap Count: 2
# Description:
#   Bug 1711244 - Integrate NativeLayerWayland with WaylandVsyncSource, r=stransky,gfx-reviewers,jrmuizel
#   
#   Make the vsync source request frame callbacks from opaque native
#   layers. This is necessary as opaque layers may occlude the
#   MozContainer surface, which is normally used for frame callbacks.
# ==============================================================================

diff -r bc5ca21bf534 -r 3c74d415fafd widget/gtk/WaylandVsyncSource.h
--- a/widget/gtk/WaylandVsyncSource.h	Thu Jun 10 14:58:53 2021 +0000
+++ b/widget/gtk/WaylandVsyncSource.h	Thu Jun 10 15:12:54 2021 +0000
@@ -6,16 +6,19 @@
 #ifndef _WaylandVsyncSource_h_
 #define _WaylandVsyncSource_h_
 
+#include "base/thread.h"
 #include "mozilla/RefPtr.h"
 #include "mozilla/Mutex.h"
 #include "mozilla/Monitor.h"
+#include "mozilla/layers/NativeLayerWayland.h"
 #include "MozContainer.h"
+#include "nsWaylandDisplay.h"
 #include "VsyncSource.h"
-#include "base/thread.h"
-#include "nsWaylandDisplay.h"
 
 namespace mozilla {
 
+using layers::NativeLayerRootWayland;
+
 /*
  * WaylandVsyncSource
  *
@@ -39,10 +42,7 @@
  */
 class WaylandVsyncSource final : public gfx::VsyncSource {
  public:
-  explicit WaylandVsyncSource(MozContainer* container) {
-    MOZ_ASSERT(NS_IsMainThread());
-    mGlobalDisplay = new WaylandDisplay(container);
-  }
+  WaylandVsyncSource() { mGlobalDisplay = new WaylandDisplay(); }
 
   virtual ~WaylandVsyncSource() { MOZ_ASSERT(NS_IsMainThread()); }
 
@@ -50,7 +50,11 @@
 
   class WaylandDisplay final : public mozilla::gfx::VsyncSource::Display {
    public:
-    explicit WaylandDisplay(MozContainer* container);
+    WaylandDisplay();
+
+    void MaybeUpdateSource(MozContainer* aContainer);
+    void MaybeUpdateSource(
+        const RefPtr<NativeLayerRootWayland>& aNativeLayerRoot);
 
     void EnableMonitor();
     void DisableMonitor();
@@ -75,13 +79,14 @@
     void ClearFrameCallback();
     void CalculateVsyncRate(TimeStamp vsyncTimestamp);
 
-    Mutex mEnabledLock;
+    Mutex mMutex;
     bool mIsShutdown;
     bool mVsyncEnabled;
     bool mMonitorEnabled;
+    bool mCallbackRequested;
     struct wl_display* mDisplay;
-    struct wl_callback* mCallback;
     MozContainer* mContainer;
+    RefPtr<NativeLayerRootWayland> mNativeLayerRoot;
     TimeDuration mVsyncRate;
     TimeStamp mLastVsyncTimeStamp;
   };